### ISSUE TITLE
Output Layer Type

### ISSUE THREAD

**Issue by ryanleary** [SeanNaren/deepspeech.pytorch#65]
I'm working on implementing a beam decoder for this, and just realized that the output values do not appear to be posteriors. In `model.py` I see the output layer is a `Linear` layer. Why not a Softmax or LogSoftmax activation? I suppose such a layer is not strictly necessary when doing a Greedy decode (and less efficient), but will be necessary for more complicated decoders. Just wondering if there's a specific reason or if there's something I'm missing.

Thanks!

**Comment by SeanNaren** [comment]
This is because warp ctc does this internally already, so we would need to do this manually for other decoders!

**Comment by ryanleary** [comment]
Right -- ok that's what I was thinking.

I am going to submit a PR that lets you specify an optional activation function on the model to be used only at inference time so as to not further slow down training. Does that seem reasonable?

**Comment by SeanNaren** [comment]
Don't mind that, just to check however would it not be possible to incorporate that into the decoder?

**Comment by ryanleary** [comment]
We could, but, at least in theory, I think a softmax or logsoftmax is used in the final layer in the DeepSpeech and DeepSpeech2 papers. The output of the network is supposed to be posterior probabilities of characters at each timestep. Doing so wouldn't really impact the current `ArgMaxDecoder`, and most decoder implementations I can think of would expect probabilities from the final layer.

**Comment by SeanNaren** [comment]
Makes sense! I think the option would be good :)

**Comment by ryanleary** [comment]
It looks like the reshaping from `SequenceWise` makes this a challenge to implement as an option. I guess the most straight forward approaches are to either:
a) add `nn.Softmax()` inside the `fully_connected` sequential definition
b) just apply the softmax function at decode time for each timestep as you suggested

I'll look more into option b as I suppose it does give the most flexibility.

**Comment by SeanNaren** [comment]
Agreed I think option B is a lot easier but again may throw people off who expect a softmax output from the model. However if we manage to add complex decoding packages with LMs and dictionaries then I feel we can hide that behind for now!

Injecting a softmax layer would be a bit tricky, mainly where abouts in the pipeline that could happen sensibly

**Comment by ryanleary** [comment]
Agreed. This can be closed I think -- perhaps adding a note to the README/wiki in the future just to highlight that there is _no_ softmax activation on the output layer would be prudent.

**Comment by SeanNaren** [comment]
Added disclaimer, thanks Ryan!
